x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -344,18 +350,101 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
                 bulkRequest,
                 TransportSingleItemBulkWriteAction.<IndexResponse>wrapBulkResponse(ActionListener.wrap(indexResponse -> {
                     assert docId.equals(indexResponse.getId());
-                    // TODO: replace with actual domain information
                     final VersionedDocument versionedDocument = new VersionedDocument(
                         profileDocument,
                         indexResponse.getPrimaryTerm(),
                         indexResponse.getSeqNo()
                     );
                     listener.onResponse(versionedDocument.toProfile(Set.of()));
-                }, listener::onFailure))
+                }, e -> {
+                    if (e instanceof VersionConflictEngineException) {
+                        // Document already exists with the specified ID, get the document with the ID
+                        // and check whether it is the right profile
+                        getVersionedDocument(profileDocument.uid(), ActionListener.wrap(versionedDocument -> {
+                            if (isTheRightProfileDocument(subject, versionedDocument.doc)) {
+                                // The profile document is a match, still need to update it with current auth info before return
+                                // listener.onResponse(versionedDocument.toProfile(Set.of()));
+                                updateProfileForActivate(subject, versionedDocument, listener);
+                            } else {
+                                // The profile document is NOT a match, this means either genuine hash collision or profile document
+                                // was manually updated, e.g. realm or user rename etc. So we attempt to differentiate from the
+                                // existing profile document by increase the differentiator number by 1.
+                                incrementDifferentiatorAndCreateNewProfile(subject, profileDocument, listener);
+                            }
+                        }, listener::onFailure));
```

Comment:
As commented above, this `getVersionedDocument` call can be changed to get all documents with UID beginning with the baseUid for efficiency.